File: dotnet/Allmark/Inline/FootnoteRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class FootnoteRule
{
	public static InlineRule Create()
	{
		return new InlineRule
		{
			Name = "footnote",
			Test = TestFootnote,
		};
	}

	private static bool TestFootnote(InlineParserState state, MarkdownNode parent)
	{
		var ch = Utils.GetChar(state.Src, state.I);

		if (!Utils.IsEscaped(state.Src, state.I))
		{
			if (ch == '[')
			{
				return TestFootnoteOpen(state, parent);
			}

			if (ch == ']')
			{
				return TestFootnoteClose(state, parent);
			}
		}

		return false;
	}

	private static bool TestFootnoteOpen(InlineParserState state, MarkdownNode parent)
	{
		var start = state.I;

		// Check for [^ pattern which indicates a footnote reference
		if (Utils.GetChar(state.Src, start + 1) != '^')
		{
			return false;
		}

		var markup = "[^";

		// Add a new text node which may turn into a footnote
		var text = Utils.NewNode("text", false, start, state.Line, 1, markup, 0);
		parent.Children!.Add(text);

		state.I += 2;
		state.Delimiters.Add(new Delimiter { Markup = markup, Start = start, Length = 2 });

		return true;
	}

	private static bool TestFootnoteClose(InlineParserState state, MarkdownNode parent)
	{
		// Find the matching footnote delimiter
		Delimiter? startDelimiter = null;
		var i = state.Delimiters.Count;
		while (i-- > 0)
		{
			var prevDelimiter = state.Delimiters[i];
			if (!prevDelimiter.Handled)
			{
				if (prevDelimiter.Markup == "[^")
				{
					startDelimiter = prevDelimiter;
					break;
				}
			}
		}

		if (startDelimiter != null)
		{
			// Convert the text node into a footnote node
			var j = parent.Children?.Count ?? 0;
			while (j-- > 0)
			{
				var lastNode = parent.Children?[j];
				if (lastNode?.Index == startDelimiter.Start)
				{
					var label = state.Src.Substring(
						startDelimiter.Start + startDelimiter.Markup.Length,
						state.I - (startDelimiter.Start + startDelimiter.Markup.Length));

					// No special characters
					if (System.Text.RegularExpressions.Regex.IsMatch(label, @"[^a-zA-Z0-9]"))
					{
						return false;
					}

					// Check for balanced brackets
					var level = 0;
					for (var k = 0; k < label.Length; k++)
					{
						if (label[k] == '\\')
						{
							k++;
						}
						else if (label[k] == '[')
						{
							level++;
						}
						else if (label[k] == ']')
						{
							level--;
						}
					}
					if (level != 0)
					{
						return false;
					}

					// Swallow anything in brackets afterwards
					// Unless it's a link reference, in which case it should be treated as a link instead
					if (Utils.GetChar(state.Src, state.I + 1) == '[')
					{
						var start = state.I + 2;
						for (var k = start; k < state.Src.Length; k++)
						{
							if (state.Src[k] == ']')
							{
								var linkRef = state.Src.Substring(start, k - start);
								linkRef = Utils.NormalizeLabel(linkRef);
								if (state.Refs.ContainsKey(linkRef))
								{
									startDelimiter.Markup = "[";
									return false;
								}
								state.I = k;
								break;
							}
						}
					}

					// Normalize the label and look it up
					label = Utils.NormalizeLabel(label);
					if (state.Footnotes.TryGetValue(label, out var footnote))
					{
						state.I++;
						startDelimiter.Handled = true;

						// Create the footnote reference node with parsed children
						lastNode.Type = "footnote";
						lastNode.Info = label;
						lastNode.Markup = $"[^{label}]";
						lastNode.Children = footnote.Content.Children;

						// Parse the footnote content for inline elements
						var tempState = new InlineParserState
						{
							Rules = state.Rules,
							Src = lastNode.Content?.TrimEnd() ?? "",
							I = 0,
							Line = lastNode.Line,
							LineStart = 0,
							Indent = 0,
							Delimiters = [],
							Refs = state.Refs,
							Footnotes = state.Footnotes,
						};
						Parse.ParseInline.Execute(tempState, lastNode);

						return true;
					}

					startDelimiter.Handled = true;
					break;
				}
			}
		}

		return false;
	}
}

File: web/src/inline/footnoteRule.ts
import parseInline from "../parse/parseInline";
import type Delimiter from "../types/Delimiter";
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import {
	BACKSLASH_CODE,
	BRACKET_OPEN_CODE,
	BRACKET_CLOSE_CODE,
	CARET_CODE,
} from "../utils/charCodes";
import newText from "../utils/newText";
import normalizeLabel from "../utils/normalizeLabel";

const rule: InlineRule = {
	name: "footnote",
	test: testFootnote,
};
export default rule;

function testFootnote(state: InlineParserState, parent: MarkdownNode): boolean {
	if (!state.isEscaped) {
		let charCode = state.src.charCodeAt(state.i);

		if (charCode === BRACKET_OPEN_CODE) {
			return testFootnoteOpen(state, parent);
		}

		if (charCode === BRACKET_CLOSE_CODE) {
			return testFootnoteClose(state, parent);
		}
	}

	return false;
}

function testFootnoteOpen(state: InlineParserState, parent: MarkdownNode) {
	let start = state.i;

	// Check for [^ pattern which indicates a footnote reference
	if (state.src.charCodeAt(start + 1) !== CARET_CODE) {
		return false;
	}

	let markup = "[^";

	// Add a new text node which may turn into a footnote
	let text = newText(state.parentIndex + start, state.line, markup, 0);
	parent.children!.push(text);

	state.i += 2;
	state.delimiters.push({ markup, start, length: 2 });

	return true;
}

function testFootnoteClose(state: InlineParserState, parent: MarkdownNode) {
	// Find the matching footnote delimiter
	let startDelimiter: Delimiter | undefined;
	let i = state.delimiters.length;
	while (i--) {
		let prevDelimiter = state.delimiters[i];
		if (!prevDelimiter.handled) {
			if (prevDelimiter.markup === "[^") {
				startDelimiter = prevDelimiter;
				break;
			}
		}
	}

	if (startDelimiter !== undefined) {
		// Convert the text node into a footnote node
		let i = parent.children!.length;
		while (i--) {
			let lastNode = parent.children![i];
			if (lastNode.index === state.parentIndex + startDelimiter.start) {
				let label = state.src.substring(
					startDelimiter.start + startDelimiter.markup.length,
					state.i,
				);

				// No special characters
				if (/[^a-zA-Z0-9]/.test(label)) {
					return false;
				}

				// Check for balanced brackets
				let level = 0;
				for (let i = 0; i < label.length; i++) {
					if (label.charCodeAt(i) === BACKSLASH_CODE) {
						i++;
					} else if (label.charCodeAt(i) === BRACKET_OPEN_CODE) {
						level++;
					} else if (label.charCodeAt(i) === BRACKET_CLOSE_CODE) {
						level--;
					}
				}
				if (level != 0) {
					return false;
				}

				// Swallow anything in brackets afterwards
				// Unless it's a link reference, in which case it should be treated as a link instead
				if (state.src.charCodeAt(state.i + 1) === BRACKET_OPEN_CODE) {
					const start = state.i + 2;
					for (let i = start; i < state.src.length; i++) {
						if (state.src.charCodeAt(i) === BRACKET_CLOSE_CODE) {
							let linkRef = state.src.substring(start, i);
							linkRef = normalizeLabel(linkRef);
							if (state.refs[linkRef] !== undefined) {
								startDelimiter.markup = "[";
								return false;
							}
							state.i = i;
							break;
						}
					}
				}

				// Normalize the label and look it up
				label = normalizeLabel(label);
				let footnote = state.footnotes[label];

				if (footnote !== undefined) {
					state.i++;
					startDelimiter.handled = true;

					// Create a footnote reference node with parsed children
					lastNode.type = "footnote";
					lastNode.info = label;
					lastNode.markup = `[^${label}]`;
					lastNode.children = footnote.content.children;
					lastNode.length = state.parentIndex + state.i - lastNode.index;

					// Parse the footnote content for inline elements
					let tempState: InlineParserState = {
						rules: state.rules,
						src: footnote.content.content,
						i: 0,
						line: lastNode.line,
						lineStart: 0,
						indent: 0,
						isEscaped: false,
						delimiters: [],
						refs: state.refs,
						footnotes: state.footnotes,
						parentIndex: lastNode.index,
					};
					parseInline(tempState, lastNode);

					return true;
				}

				startDelimiter.handled = true;
				break;
			}
		}
	}

	return false;
}
